ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -0,0 +1,200 @@
+
+    const size_t physical_area = physical_size.height() * physical_size.width();
+    const size_t page_area = page_shape.height() * page_shape.width();
+
+    const size_t page_count = physical_area / page_area;
+    const size_t page_size_bytes = compute_page_size_bytes(page_shape);
+
+    return page_count * page_size_bytes;
+}
+
+size_t TensorLayout::compute_page_size_bytes(const ttnn::SimpleShape& shape) const {
+    const auto physical_size = compute_physical_shape(shape);
+    const auto page_shape = compute_page_shape(physical_size);
+    return compute_page_size_bytes(page_shape);
+}
+
+size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
+    return m_page_config.get_page_size_bytes(page_size, m_dtype);
+}
+
+Size TensorLayout::compute_physical_shape(const ttnn::SimpleShape& shape) const {
+    const int rank = static_cast<int>(shape.rank());
+    const int alignment_rank = static_cast<int>(m_alignment.size());
+    const int max_rank = std::max(rank, alignment_rank);
+    size_t width = 1;
+    size_t height = 1;
+
+    // Iterate dims in reverse order and ensure alignment
+    // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
+    for (int i = -1; i >= -max_rank; --i) {
+        auto& dim = i == -1 ? width : height;
+        if(i >= -rank) {
+            dim *= shape[i];
+        }
+
+        // Align the current dimension if alignment is available
+        if (i >= -alignment_rank) {
+            dim = round_up(dim, m_alignment[i]);
+        }
+    }
+
+    Size size{height, width};
+    return size;
+}
+
+Size TensorLayout::compute_page_shape(const Size& physical_size) const {
+    if(m_memory_config.memory_layout == TensorMemoryLayout::SINGLE_BANK) {
+        return physical_size;
+    }
+
+    return m_page_config.get_page_shape(physical_size, m_memory_config);
+}
+
+Strides TensorLayout::compute_strides(const ttnn::SimpleShape& shape) const {
+    const int rank = static_cast<int>(shape.rank());
+    const int alignment_rank = static_cast<int>(m_alignment.size());
+
+    Strides strides(rank, 1);
+    for (int i = rank - 2; i >= 0; i--) {
+        strides[i] = strides[i + 1] * shape[i + 1];
+
+        const int alignment_index = i - (rank - alignment_rank) + 1;
+        if(alignment_index >= 0) {
+            strides[i] = round_up(strides[i], m_alignment[alignment_index]);
+        }
+    }
+
+    return strides;
+}
+
+ttnn::SimpleShape TensorLayout::compute_padded_shape(const ttnn::SimpleShape& shape) const
+{
+    ttnn::SmallVector<uint32_t> padded_shape(shape.rank());
+    int rank_index = static_cast<int>(shape.rank()) - 1;
+    int alignment_index = static_cast<int>(m_alignment.size()) - 1;
+    size_t accum_alignment = 1;
+
+    for (;rank_index >= 0 && alignment_index >= 0; rank_index--, alignment_index--) {
+        // The last 2 dimensions of a shape are special
```

Comment:
these represent batch? if so then a comment explaining that would be helpful

Reply:
No, these are last innermost dimensions. Special because they might be tile aligned.
CC @sminakov-tt